### PR TITLE
Support farbfeld

### PR DESCRIPTION
**Pull request by nabijaczleweli**
Also one warning (among the dozens unrelated ones spewed by cargo) and PNG doccomment.

I license past and future contributions under the dual MIT/Apache-2.0 license,
allowing licensees to chose either at their option.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -40,7 +40,7 @@ quickcheck = "0.9"
 criterion = "0.3"
 
 [features]
-default = ["gif", "jpeg", "ico", "png", "pnm", "tga", "tiff", "webp", "bmp", "hdr", "dxt", "dds", "jpeg_rayon"]
+default = ["gif", "jpeg", "ico", "png", "pnm", "tga", "tiff", "webp", "bmp", "hdr", "dxt", "dds", "farbfeld", "jpeg_rayon"]
```
Comment (HeroicKatora) [inline Cargo.toml]: Please add an extra configuration with _only_ the new features to `.travis.yml`.

File: src/farbfeld.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        })
+    }
+}
+
+impl<R: Read> Read for FarbfeldReader<R> {
+    fn read(&mut self, mut buf: &mut [u8]) -> io::Result<usize> {
+        // Note: this truncates odd buf lengths to the nearest-lower multiple of two
+        let channels_to_read = buf.len() / 2;
+
+        for _ in 0..channels_to_read {
```
[the remark this answers is not in the document]
Comment (nabijaczleweli) [reply src/farbfeld.rs]: Done

File: Cargo.toml (hunk 1 above)
Comment (nabijaczleweli) [reply Cargo.toml]: Done

File: src/farbfeld.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            height: read_dimm(&mut inner)?,
+            inner,
+        })
+    }
+}
+
+impl<R: Read> Read for FarbfeldReader<R> {
+    fn read(&mut self, mut buf: &mut [u8]) -> io::Result<usize> {
+        // Note: this truncates odd buf lengths to the nearest-lower multiple of two
+        let channels_to_read = buf.len() / 2;
```
[the remark this answers is not in the document]
Comment (nabijaczleweli) [reply src/farbfeld.rs]: Dome

**Comment by nabijaczleweli** [comment]
`ImageDecoderExt` is entirely undocumented and I didn't see any tests, so I couldn't really test it beside copying another implementation and making sure it builds.

**Comment by HeroicKatora** [comment]
@fintelia I must admit that I hadn't heard of the format before as well but generally, `suckless.org` creates pretty stable and simple products, it truly sucks less. Since it doesn't add new dependencies and doesn't influence other interfaces, why not add it? Maybe it could stay an optional feature instead of a default feature for now.

With regards to `ImageDecoderExt`, I'd do that in a follow-up PR. Especially if the changes here are very clear and simple but the additional changes would require some discussion and mentoring it stalls finished work.

**Comment by nabijaczleweli** [comment]
Right, moved the `ImageDecoderExt` commit to `farbfeld-followup` and removed `"farbfeld"` from the `features.default` in Cargo.toml.

**Comment by HeroicKatora** [comment]
Looking forward to the follow up :+1:

**Comment by fintelia** [comment]
@nabijaczleweli the ImageDecoderExt trait is documented [here](https://docs.rs/image/0.23.2/image/trait.ImageDecoderExt.html). If it isn't clear from that page what the trait is supposed to do/how to test it, could you open an issue and we can see about improving things?

**Comment by nabijaczleweli** [comment]
@HeroicKatora: #1173 
@fintelia: #1174

**Pull request merged**